zilliqa/src/consensus.rs

```diff
@@ -2627,29 +2627,35 @@ impl Consensus {
             trace!("applying {} transactions to state", transactions.len());
         }
 
-        let transactions: Result<Vec<_>> = transactions.into_iter().map(|tx| tx.verify()).collect();
-        let mut transactions = transactions?;
-
-        let transaction_hashes = transactions
-            .iter()
-            .map(|tx| format!("{:?}", tx.hash))
-            .join(",");
+        let mut verified_txns = Vec::new();
 
         // We re-inject any missing Intershard transactions (or really, any missing
-        // transactions) from our mempool. If any txs are unavailable either in the
+        // transactions) from our mempool. If any txs are unavailable in both the
         // message or locally, the proposal cannot be applied
         for (idx, tx_hash) in block.transactions.iter().enumerate() {
-            if transactions.get(idx).is_some_and(|tx| tx.hash == *tx_hash) {
-                // all good
-            } else {
-                let Some(local_tx) = self.transaction_pool.pop_transaction(*tx_hash) else {
-                    warn!("Proposal {} at view {} referenced a transaction {} that was neither included in the broadcast nor found locally - cannot apply block", block.hash(), block.view(), tx_hash);
-                    return Ok(());
-                };
-                transactions.insert(idx, local_tx);
-            }
+            // Prefer to insert verified txn from pool. This is faster.
+            let txn = match self.transaction_pool.pop_transaction(*tx_hash) {
+                Some(txn) if txn.hash == *tx_hash => txn,
+                _ => match transactions
+                    .get(idx)
+                    .and_then(|sig_tx| sig_tx.clone().verify().ok())
+                {
+                    // Otherwise, recover txn from proposal. This is slower.
+                    Some(txn) if txn.hash == *tx_hash => txn,
+                    _ => {
+                        warn!("Proposal {} at view {} referenced a transaction {} that was neither included in the broadcast nor found locally - cannot apply block", block.hash(), block.view(), tx_hash);
+                        return Ok(());
+                    }
+                },
+            };
+            verified_txns.insert(idx, txn);
```

Comment:
If I understand correctly, its always true that `idx == verified_txns.len()` (i.e. we're always adding to the end of the list) which makes this equivalent to `verified_txns.push(txn);`. Can we just push and remove the `idx` to make things clearer?